[libc++] LWG3187: P0591R4 reverted DR 2586 fixes to `scoped_allocator_adaptor::construct()`

frederick-vs-ja opened the pull request:
Applying it (adding missing `const`) to pre-C++20 dispatching mechanisms.

Also
- make dispatch mechanisms in pre-C++20 `construct` more consistent,
- add fallback dispatching and overloads to emit better error messages, and
- partially implement LWG4312.

Fixes #100256.

frederick-vs-ja commented on libcxx/include/scoped_allocator:
```diff
@@ -421,23 +424,30 @@ public:
+      allocator_traits<typename _OM::type>::construct(_OM()(outer_allocator()), __p, std::forward<_Args>(__args)...);
+    } else if constexpr (is_constructible<_Type, allocator_arg_t, const inner_allocator_type&, _Args...>::value) {
+      allocator_traits<typename _OM::type>::construct(
+          _OM()(outer_allocator()), __p, allocator_arg, inner_allocator(), std::forward<_Args>(__args)...);
```
I just realized that `inner_allocator()` is non-const here. Perhaps there should be a similar evil test for `scoped_allocator_adaptor`.

frederick-vs-ja replied:
It seems sufficient to modify existing tests.

philnik777 reviewed (changes requested):
Shouldn't we be able to use the utilities from `uses_allocator_construction.h`?

frederick-vs-ja commented:
> Shouldn't we be able to use the utilities from `uses_allocator_construction.h`?

I found that there're some behavior differences introduced by [P0591R4](https://wg21.link/p0591r4) and I'm not sure whether the C++20 behavior should be backported.
- The mechanisms for P0591R4 recursively handle nested `pair<pair<...>, ...>` while pre-C++20 ones didn't, which was already noticed in P0591R4.
- There're small type changes around `allocator_arg_t` and `piecewise_construct_t` (from `const&` to `&&`). This generally doesn't matter in practice but is detectable.

philnik777 commented:
> > I think we should rather modify the utilities to have the required behaviour in C++17.
> 
> Oh, for `scoped_allocator_adaptor` we need the mechanisms working in C++11 mode. It would be quite non-trivial to rework them for C++11, as (at least) Clang doesn't seem to accept using deduced return types in C++11 mode.

Urgh. WDYT about implementing the `polymorphic_allocator` part first and then we can see how big the diff would be for C++11?

frederick-vs-ja committed:
[libc++] LWG3187: P0591R4 reverted DR 2586 fixes to `scoped_allocator_adaptor::construct()`

Applying it (adding missing `const``) to pre-C++20 dispatching
mechanisms.

Also
- make dispatch mechanisms in pre-C++20 `construct` more consistent,
- add fallback dispatching and overloads to emit better error messages,
and
- partially implement LWG4312.
